Consolidate steps

sunnytownsend opened the pull request:
- reduce number of workflow functions to 5 higher level steps - that might be sensible chunks of code for swopping in and out in future more flexible versions of the model
- update workflow and catalog accordingly
- make changes to the model so that it adheres to [this flowchart](https://docs.google.com/drawings/d/1u1hvzacgl7uuatJe4mV1nLzp6xp8zf41rc7HFU_v020/edit)
- check the `defmodel` is valid
- update the test ns to keep unit tests for fns still in use

sunnytownsend committed:
New model structure - consolidate to 6 high level steps

Modify workflow, catalog, model fns and library

Commit:
Fix defmodel, test data, calc fns and tests

sunnytownsend commented on data/testing_data.edn:
```diff
@@ -83,68 +86,102 @@
+                        {:gss-code "001" :age-group :35-39 :sex "M"
+                         :year 2015 :relationship "C" :resident-popn 9929.27864215}]
+
+  :witan-institutional-popn [{:gss-code "001" :age-group :30-34 :sex "F"
```
you are not testing the age 75+ different calcs for inst popn - do you have another test for those?

sunnytownsend commented on data/testing_data.edn:
```diff
@@ -83,68 +86,102 @@
 
   :total-households [{:gss-code "001" :year 2015 :households 14000.0}]
 
   :occupancy-rates [{:gss-code "001" :year 2015 :occupancy-rates 0.313}]
```
remove `occupancy-rates`

sunnytownsend commented on data/testing_data.edn:
```diff
@@ -83,68 +86,102 @@
 
   :household-popn [{:gss-code "001" :age-group :30-34 :sex "F"
-                    :year 2015  :relationship "S" :household-popn 5714.285714285714}
+                    :year 2015  :relationship "S" :household-popn 5606.060606061}
```
these numbers are much smaller than for the `input-dclg-household-popn` which I am surprised to see - how did you get these numbers?

sunnytownsend commented on test/witan/models/household_test.clj:
```diff
@@ -91,49 +91,96 @@
+      (is (= res-popn correct-output)))))
+
 (deftest sum-resident-popn-test
   (testing "The resident projections are summed by household type"
```
replace 'household type' with 'relationship'

sunnytownsend commented on src/witan/models/household.clj:
```diff
@@ -15,176 +15,219 @@
+
+(defn sum-resident-popn
   "Takes in the resident populations.
    Returns the same population summed by household type."
```
replace 'household type' with 'relationship'

Commit:
Implement @sunnytownsend 's comments

Rm occupancy rates, test data + test for 75+, household type -> relationship
update convert-to-dwellings, rename fn

sunnytownsend commented on data/testing_data.edn:
```diff
@@ -1,80 +1,96 @@
+                                              {:gss-code "001" :year 2015 :sex "M" :relationship "C"
+                                               :age-group :35-39 :hh-repr-rates 0.3}]
+
+  :input-dclg-dwellings [{:gss-code "001" :year 2015 :dwellings 100.0}
```
we will only get one number for dwellings

Eleonore9 replied:
Oh yes, sure! It's the value available for the latest year

sunnytownsend commented on data/testing_data.edn:
```diff
@@ -1,80 +1,96 @@
+  :input-dclg-dwellings [{:gss-code "001" :year 2015 :dwellings 100.0}
+                         {:gss-code "001" :year 2016 :dwellings 100.0}]
+
+  :input-vacancy-dwellings [{:gss-code "001" :year 2015 :vacancy-dwellings 16.0}
```
we will only get one number for vacancy dwellings too

Eleonore9 replied:
ok, will change that now!

sunnytownsend commented on src/witan/models/household.clj:
```diff
@@ -15,176 +15,219 @@
-                  (ds/select-columns [:gss-code :year :dwellings]))})
+  [{:keys [total-households dclg-dwellings vacancy-dwellings]} {:keys [second-home-rate]}]
+  (let [vacancy-rates (-> dclg-dwellings
+                          (wds/join vacancy-dwellings [:gss-code :year])
```
so here we need to pull out the projection years from elsewhere e.g. witan popn input, and use those

sunnytownsend commented on src/witan/models/household.clj:
```diff
@@ -15,176 +15,219 @@
+  (let [popn-by-5yrs-bands (grp-popn-proj population)
+        dclg-resident-popn (create-resident-popn dclg-household-popn dclg-institutional-popn)
+        dclg-resident-by-relationship (sum-resident-popn dclg-resident-popn)]
+    {:resident-popn (combine-resident-proj dclg-resident-popn
```
suggest change `combine` to `calc`

sunnytownsend commented on src/witan/models/household.clj:
```diff
@@ -15,176 +15,219 @@
+                                                               [:75-79 :80-84 :85+])
+                                                       (* (/ dinst dres) res) dinst)))
+                           (ds/select-columns [:gss-code :age-group :sex :year
+                                               :relationship :institutional-popn]))})
```
very nice!

acron0 commented on src/witan/models/household.clj:
```diff
@@ -15,176 +15,232 @@
 
-(definput get-resident-popn-1-0-0
-  {:witan/name :hh-model/get-resident-popn
+(definput get-dclg-household-popn-1-0-0
```
I kinda prefer the 'in' prefix, because 'get' is a verb and these aren't functions per se.

sunnytownsend replied:
we actually got rid of the prefix for the CCM in the end - and it meant the flowchart and model could match, so could do that here too

acron0 commented on src/witan/models/household.clj:
```diff
@@ -15,176 +15,232 @@
+(defworkflowfn apportion-popn-by-relationship-1-0-0
+  "Takes in a population (output of CCM), dclg household and institutional
+   populations. Returns a resident population with relationships and grouped
+   by five years bands."
```
Will it always be 5 year bands?

sunnytownsend replied:
with this model and dclg data - yes. with a different household model or a different country, poss not. Will change to be clearer - that the age bands are derived from the dclg data (whatever banding that is)

acron0 commented on src/witan/models/household.clj:
```diff
@@ -15,176 +15,232 @@
+                    (wds/join total-households [:gss-code :year])
+                    (wds/add-derived-column :dwellings
+                                            [:households :second-home-rates :vacancy-rates]
+                                            (fn [hh shr vr] (/ hh (- 1 (- vr shr)))))
```
`dec`

acron0 reviewed (commented):
Only really minor things. Also, it'd be good if all your different tests weren't in one file.

Commit:
Implement @acron0 's comments

Split tests in several ns, rename definput for consistency
Update docstring and fn

acron0 commented on src/witan/models/household.clj:
```diff
@@ -217,7 +217,7 @@
                     (wds/add-derived-column :dwellings
                                             [:households :second-home-rates :vacancy-rates]
-                                            (fn [hh shr vr] (/ hh (- 1 (- vr shr)))))
+                                            (fn [hh shr vr] (/ hh (- 1 vr shr))))
```
This is fine but be aware it's technically slower

acron0 reviewed (approved):
I can still see a var called `popn-by-5yrs-bands` but other than that this is good

Commit:
Add testing data to test convert-to-dwellings fn

Commit:
More changes from @acron0's comments

Commit:
Improve calc
